From e41efce241de75a9f6d4668c455603f2d5bd8744 Mon Sep 17 00:00:00 2001
From: Debarshi Ray <debarshir@gnome.org>
Date: Tue, 17 May 2016 21:08:55 +0200
Subject: [PATCH] Restore separate menuitems for opening tabs and windows

Without the separate menuitems, it was not possible to open new tabs or
windows solely by using the mouse if the preference was set to windows
or tabs respectively. This isn't ideal for accessibility.

The preference is also hard to discover and there was simply no strong
justification for unifying the menuitems in the first place.

The org.gnome.Terminal.Legacy.Settings:new-terminal-mode preference is
now ignored.

This effectively reverts 99fc0136a5be6323b81b8b339482bc699b53e1f9 and
1fecaa3685a8a4d4e7027fc6c9e60df2ca69f5b6

https://bugzilla.redhat.com/show_bug.cgi?id=1300826
---
 src/preferences.ui      | 58 -------------------------------------------------
 src/terminal-appmenu.ui |  7 ------
 src/terminal-prefs.c    |  9 +-------
 src/terminal-schemas.h  |  1 -
 src/terminal-window.c   | 49 +++++++++++++++++++++--------------------
 src/terminal.xml        |  7 ++++--
 6 files changed, 31 insertions(+), 100 deletions(-)

diff --git a/src/preferences.ui b/src/preferences.ui
index df825088f3bd..bb4d362fd233 100644
--- a/src/preferences.ui
+++ b/src/preferences.ui
@@ -1,23 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <interface>
   <!-- interface-requires gtk+ 3.0 -->
-  <object class="GtkListStore" id="new-terminal-mode-liststore">
-    <columns>
-      <!-- column-name gchararray -->
-      <column type="gchararray"/>
-      <column type="gchararray"/>
-    </columns>
-    <data>
-      <row>
-        <col id="0" translatable="yes" comments="Open new terminal in new window">Window</col>
-        <col id="1" translatable="no">window</col>
-      </row>
-      <row>
-        <col id="0" translatable="yes" comments="Open new terminal in new tab">Tab</col>
-        <col id="1" translatable="no">tab</col>
-      </row>
-    </data>
-  </object>
   <object class="GtkDialog" id="preferences-dialog">
     <property name="can_focus">False</property>
     <property name="border_width">5</property>
@@ -154,47 +137,6 @@
                     <property name="position">3</property>
                   </packing>
                 </child>
-                <child>
-                  <object class="GtkBox" id="hbox140">
-                    <property name="visible">True</property>
-                    <property name="can_focus">False</property>
-                    <property name="spacing">12</property>
-                    <property name="orientation">horizontal</property>
-                    <child>
-                      <object class="GtkLabel" id="label480">
-                        <property name="visible">True</property>
-                        <property name="can_focus">False</property>
-                        <property name="label" translatable="yes">Open _new terminals in:</property>
-                        <property name="use_underline">True</property>
-                        <property name="mnemonic_widget">new-terminal-mode-combobox</property>
-                      </object>
-                      <packing>
-                        <property name="expand">False</property>
-                        <property name="fill">False</property>
-                        <property name="position">0</property>
-                      </packing>
-                    </child>
-                    <child>
-                      <object class="GtkComboBox" id="new-terminal-mode-combobox">
-                        <property name="visible">True</property>
-                        <property name="can_focus">False</property>
-                        <property name="model">new-terminal-mode-liststore</property>
-                        <property name="id-column">1</property>
-                        <child>
-                          <object class="GtkCellRendererText" id="renderer1"/>
-                          <attributes>
-                            <attribute name="text">0</attribute>
-                          </attributes>
-                        </child>
-                      </object>
-                      <packing>
-                        <property name="expand">False</property>
-                        <property name="fill">True</property>
-                        <property name="position">1</property>
-                      </packing>
-                    </child>
-                  </object>
-                </child>
               </object>
             </child>
             <child type="tab">
diff --git a/src/terminal-appmenu.ui b/src/terminal-appmenu.ui
index 01ff5a85f7cb..0c6b86b0be90 100644
--- a/src/terminal-appmenu.ui
+++ b/src/terminal-appmenu.ui
@@ -19,13 +19,6 @@
   <menu id="appmenu">
     <section>
       <item>
-        <attribute name="label" translatable="yes">_New Terminal</attribute>
-        <attribute name="action">win.new-terminal</attribute>
-        <attribute name="target" type="(ss)">('default','default')</attribute>
-      </item>
-    </section>
-    <section>
-      <item>
         <attribute name="label" translatable="yes">_Preferences</attribute>
         <attribute name="action">app.preferences</attribute>
       </item>
diff --git a/src/terminal-prefs.c b/src/terminal-prefs.c
index a57dde665b44..18eef25db042 100644
--- a/src/terminal-prefs.c
+++ b/src/terminal-prefs.c
@@ -562,7 +562,7 @@ terminal_prefs_show_preferences (GtkWindow *transient_parent,
   GtkWidget *show_menubar_button, *disable_mnemonics_button, *disable_menu_accel_button;
   GtkWidget *disable_shortcuts_button;
   GtkWidget *tree_view_container, *new_button, *edit_button, *clone_button, *remove_button;
-  GtkWidget *dark_theme_button, *new_terminal_mode_combo;
+  GtkWidget *dark_theme_button;
   GtkWidget *default_hbox, *default_label;
   GtkTreeSelection *selection;
   GSettings *settings;
@@ -584,7 +584,6 @@ terminal_prefs_show_preferences (GtkWindow *transient_parent,
                                        "preferences-dialog", &dialog,
                                        "default-show-menubar-checkbutton", &show_menubar_button,
                                        "dark-theme-checkbutton", &dark_theme_button,
-                                       "new-terminal-mode-combobox", &new_terminal_mode_combo,
                                        "disable-mnemonics-checkbutton", &disable_mnemonics_button,
                                        "disable-shortcuts-checkbutton", &disable_shortcuts_button,
                                        "disable-menu-accel-checkbutton", &disable_menu_accel_button,
@@ -619,12 +618,6 @@ terminal_prefs_show_preferences (GtkWindow *transient_parent,
                    "active",
                    G_SETTINGS_BIND_GET | G_SETTINGS_BIND_SET);
 
-  g_settings_bind (settings,
-                   TERMINAL_SETTING_NEW_TERMINAL_MODE_KEY,
-                   new_terminal_mode_combo,
-                   "active-id",
-                   G_SETTINGS_BIND_GET | G_SETTINGS_BIND_SET);
-
   /* Keybindings tab */
 
   g_settings_bind (settings,
diff --git a/src/terminal-schemas.h b/src/terminal-schemas.h
index 42d542f7ed34..24c468af9630 100644
--- a/src/terminal-schemas.h
+++ b/src/terminal-schemas.h
@@ -75,7 +75,6 @@ G_BEGIN_DECLS
 #define TERMINAL_SETTING_ENABLE_MNEMONICS_KEY           "mnemonics-enabled"
 #define TERMINAL_SETTING_ENABLE_SHORTCUTS_KEY           "shortcuts-enabled"
 #define TERMINAL_SETTING_ENCODINGS_KEY                  "encodings"
-#define TERMINAL_SETTING_NEW_TERMINAL_MODE_KEY          "new-terminal-mode"
 #define TERMINAL_SETTING_SCHEMA_VERSION                 "schema-version"
 #define TERMINAL_SETTING_SHELL_INTEGRATION_KEY          "shell-integration-enabled"
 
diff --git a/src/terminal-window.c b/src/terminal-window.c
index 8d9e9978aab2..8534b0015763 100644
--- a/src/terminal-window.c
+++ b/src/terminal-window.c
@@ -91,7 +91,8 @@ struct _TerminalWindowPrivate
 
 #define PROFILE_DATA_KEY "GT::Profile"
 
-#define FILE_NEW_TERMINAL_UI_PATH         "/menubar/File/FileNewTerminalProfiles"
+#define FILE_NEW_TERMINAL_TAB_UI_PATH     "/menubar/File/FileNewTabProfiles"
+#define FILE_NEW_TERMINAL_WINDOW_UI_PATH  "/menubar/File/FileNewWindowProfiles"
 #define SET_ENCODING_UI_PATH              "/menubar/Terminal/TerminalSetEncoding/EncodingsPH"
 #define SET_ENCODING_ACTION_NAME_PREFIX   "TerminalSetEncoding"
 
@@ -369,7 +370,6 @@ action_new_terminal_cb (GSimpleAction *action,
   gs_free char *new_working_directory = NULL;
   const char *mode_str, *uuid_str;
   TerminalNewTerminalMode mode;
-  GdkModifierType modifiers;
 
   g_assert (TERMINAL_IS_WINDOW (window));
 
@@ -381,18 +381,8 @@ action_new_terminal_cb (GSimpleAction *action,
     mode = TERMINAL_NEW_TERMINAL_MODE_TAB;
   else if (g_str_equal (mode_str, "window"))
     mode = TERMINAL_NEW_TERMINAL_MODE_WINDOW;
-  else {
-    mode = g_settings_get_enum (terminal_app_get_global_settings (app),
-                                TERMINAL_SETTING_NEW_TERMINAL_MODE_KEY);
-    if (gtk_get_current_event_state (&modifiers) &&
-        (modifiers & gtk_accelerator_get_default_mod_mask () & GDK_CONTROL_MASK)) {
-      /* Invert */
-      if (mode == TERMINAL_NEW_TERMINAL_MODE_WINDOW)
-        mode = TERMINAL_NEW_TERMINAL_MODE_TAB;
-      else
-        mode = TERMINAL_NEW_TERMINAL_MODE_WINDOW;
-    }
-  }
+  else
+    return;
 
   profiles_list = terminal_app_get_profiles_list (app);
   if (g_str_equal (uuid_str, "current"))
@@ -435,9 +425,9 @@ file_new_terminal_callback (GtkAction *action,
     uuid = g_strdup ("current");
 
   name = gtk_action_get_name (action);
-  if (g_str_has_prefix (name, "FileNewTab"))
+  if (g_str_has_prefix (name, "FileNewTab") || g_str_has_prefix (name, "PopupNewTab"))
     param = g_variant_new ("(ss)", "tab", uuid);
-  else if (g_str_has_prefix (name, "FileNewWindow"))
+  else if (g_str_has_prefix (name, "FileNewWindow") || g_str_has_prefix (name, "PopupNewTerminal"))
     param = g_variant_new ("(ss)", "window", uuid);
   else
     param = g_variant_new ("(ss)", "default", uuid);
@@ -1513,8 +1503,6 @@ terminal_window_update_new_terminal_menus (TerminalWindow *window)
   gtk_action_set_visible (action, have_single_profile);
   action = gtk_action_group_get_action (priv->action_group, "FileNewWindow");
   gtk_action_set_visible (action, have_single_profile);
-  action = gtk_action_group_get_action (priv->action_group, "FileNewTerminal");
-  gtk_action_set_visible (action, have_single_profile);
 
   if (have_single_profile)
     {
@@ -1536,7 +1524,19 @@ terminal_window_update_new_terminal_menus (TerminalWindow *window)
       GSettings *profile = (GSettings *) p->data;
       char name[32];
 
-      g_snprintf (name, sizeof (name), "FileNewTerminal.%u", n);
+      g_snprintf (name, sizeof (name), "FileNewTab.%u", n);
+      terminal_window_create_new_terminal_action (window,
+                                                  profile,
+                                                  name,
+                                                  n,
+                                                  G_CALLBACK (file_new_terminal_callback));
+
+      gtk_ui_manager_add_ui (priv->ui_manager, priv->new_terminal_ui_id,
+                             FILE_NEW_TERMINAL_TAB_UI_PATH,
+                             name, name,
+                             GTK_UI_MANAGER_MENUITEM, FALSE);
+
+      g_snprintf (name, sizeof (name), "FileNewWindow.%u", n);
       terminal_window_create_new_terminal_action (window,
                                                   profile,
                                                   name,
@@ -1544,7 +1544,7 @@ terminal_window_update_new_terminal_menus (TerminalWindow *window)
                                                   G_CALLBACK (file_new_terminal_callback));
 
       gtk_ui_manager_add_ui (priv->ui_manager, priv->new_terminal_ui_id,
-                             FILE_NEW_TERMINAL_UI_PATH,
+                             FILE_NEW_TERMINAL_WINDOW_UI_PATH,
                              name, name,
                              GTK_UI_MANAGER_MENUITEM, FALSE);
 
@@ -2470,7 +2470,8 @@ terminal_window_init (TerminalWindow *window)
     {
       /* Toplevel */
       { "File", NULL, N_("_File") },
-      { "FileNewTerminalProfiles", STOCK_NEW_WINDOW, N_("Open _Terminal")},
+      { "FileNewWindowProfiles", STOCK_NEW_WINDOW, N_("Open _Terminal")},
+      { "FileNewTabProfiles", STOCK_NEW_TAB, N_("Open Ta_b")},
       { "Edit", NULL, N_("_Edit") },
       { "View", NULL, N_("_View") },
       { "Search", NULL, N_("_Search") },
@@ -2487,9 +2488,6 @@ terminal_window_init (TerminalWindow *window)
       { "FileNewTab", STOCK_NEW_TAB, N_("Open Ta_b"), "<shift><control>T",
         NULL,
         G_CALLBACK (file_new_terminal_callback) },
-      { "FileNewTerminal", STOCK_NEW_TAB, N_("Open _Terminal"), NULL,
-        NULL,
-        G_CALLBACK (file_new_terminal_callback) },
       { "FileNewProfile", "document-open", N_("New _Profile"), "",
         NULL,
         G_CALLBACK (file_new_profile_callback) },
@@ -2636,6 +2634,9 @@ terminal_window_init (TerminalWindow *window)
       { "PopupNewTerminal", NULL, N_("Open _Terminal"), NULL,
         NULL,
         G_CALLBACK (file_new_terminal_callback) },
+      { "PopupNewTab", NULL, N_("Open Ta_b"), NULL,
+        NULL,
+        G_CALLBACK (file_new_terminal_callback) },
       { "PopupLeaveFullscreen", NULL, N_("L_eave Full Screen"), NULL,
         NULL,
         G_CALLBACK (popup_leave_fullscreen_callback) },
diff --git a/src/terminal.xml b/src/terminal.xml
index e07e9c2dfeb4..a4933e4a919b 100644
--- a/src/terminal.xml
+++ b/src/terminal.xml
@@ -1,8 +1,10 @@
 <ui>
   <menubar>
     <menu action="File">
-      <menuitem action="FileNewTerminal" />
-      <menu action="FileNewTerminalProfiles" />
+      <menuitem action="FileNewWindow" />
+      <menu action="FileNewWindowProfiles" />
+      <menuitem action="FileNewTab" />
+      <menu action="FileNewTabProfiles" />
       <separator />
       <menuitem action="FileNewProfile" />
       <menuitem action="FileSaveContents" />
@@ -85,6 +87,7 @@
     <menuitem action="PopupCopyLinkAddress" />
     <separator />
     <menuitem action="PopupNewTerminal" />
+    <menuitem action="PopupNewTab" />
     <separator />
     <menuitem action="PopupCopy" />
     <menuitem action="PopupPaste" />
-- 
2.5.5

